### PR TITLE
Rework to be dependant on PSim rather than lin_old

### PR DESCRIPTION
**Pull request by kkrol27**
All branches will need to rebased off of master in order to compile. The older version of the lin library will be removed from Github once this PR goes through.

### CODE REVIEW

File: src/ssa.cpp [hunk 1]
```diff
@@ -93,8 +93,8 @@ unsigned char calculate_sun_vector(lin::Vector3f &sun_vec) {
   // Prepare least squares problem
```
Comment (tanishqaggarwal) [inline src/ssa.cpp]: Not super sure what's going on in these changes but it just looks like renamings

File: src/mtr.cpp [hunk 2]
```diff
@@ -51,7 +51,7 @@ void setup() {
 
 void actuate(unsigned char mtr_mode, lin::Vector3f mtr_cmd, float mtr_lim) {
 #ifdef DEBUG
-  lin::Vector<int, 3> pwm_value = lin::zeros<int, 3, 1>();
```
Comment (tanishqaggarwal) [inline src/mtr.cpp]: Why the switch to std::array?

File: platformio.ini [hunk 3]
```diff
@@ -4,15 +4,16 @@ board = teensy35
...
     https://github.com/pathfinder-for-autonomous-navigation/CommonSoftware
     https://github.com/pathfinder-for-autonomous-navigation/ChRt
     https://github.com/bblanchon/ArduinoJson
 build_flags =
-    -std=c++11
+    -std=c++14
     -Wall
     -Werror
-    -DNDEBUG ; Disable asserts on teensy
+    -DLIN_RANDOM_SEED=6828031
```
Comment (tanishqaggarwal) [inline platformio.ini]: Does this have to be a compiler flag? I'm wondering if it'd be better if we could specify seeds at runtime, so that a unit test could try many different seeds at once.

File: src/ssa.cpp (hunk 1 above)
Comment (kkrol27) [reply src/ssa.cpp]: It's just renaming for now. This will actually most likely get pulled out into the PSim repo at some point.

File: src/mtr.cpp (hunk 2 above)
Comment (kkrol27) [reply src/mtr.cpp]: `lin::internal::Tensor<...>` doesn't take integral types, only floats and doubles. This is because not all operations are supported for integral types (division can be weird, norms, factorization, etc.). I could just enable those operations specifically but that would end up being a lot of operations too keep track of.

File: platformio.ini (hunk 3 above)
Comment (kkrol27) [reply platformio.ini]: You can specify your own seeds at runtime by not using the default random `float`/`double` generator. See around this line: https://github.com/kkrol27/lin/blob/10d5722cf3285e59a44ff083b21a8527ec6b3a28/include/core/tensor/generators.hpp#L58

If you wanted to write a test case with a custom flag simply do something like this:

    TEST(..., ...) {
      lin::internal::RandomsGenerator generator(<seed>);
      lin::Vector3f v = lin::rands<lin::Vector3f>(3, 1, generator);
      ...
    }

The only downside is you can't let the call to `rands` default your sizing parameters for the returned stream (i.e. the `3, 1` for three rows and one column). The `RandomsGenerator` when used as shown above is full `constexpr` construction and use so there is essentially no overhead.

Comment (tanishqaggarwal) [reply platformio.ini]: Ah, I see. I think this makes sense but is at the expense of understandability, since the seed is physically far away from where it's actually used. Do you think it's worth it to also explicitly specify the seed in the generator?

Comment (kkrol27) [reply platformio.ini]: I guess I don't really know what you mean by "elso explicitly specify the seed in the generator"? The compiler flag is just to let people be lazy and not have to worry about specifying the random number generator every single time they want a random vector/matrix etc. Outside of test cases, I don't see a use case where we need to set the seed explicitly in software.

I'm looking into populating that flag with some plaformio build argument or something (https://docs.platformio.org/en/latest/envvars.html)

Comment (tanishqaggarwal) [reply platformio.ini]: What I mean is that rather than a test writer assuming there's some magic that specifies a seed, I think it'd be better practice for a test writer to specify a seed for the generator each time they want to use it.

Comment (tanishqaggarwal) [reply platformio.ini]: This wouldn't involve changing lin, but would just be a stylistic guideline for writing unit tests.

Comment (kkrol27) [reply platformio.ini]: Oh oh yes no I agree for sure. Especially for unit test cases where weird corner cases can result in failures. One last note on this subject just so you are aware, on a desktop the seed will auto populate with a call to `time(nullptr)` if you don't specify it in the compiler flags. This will only affect PSim stuff though.
